### ISSUE TITLE
[Proposal]: IOnceEnumerable

### ISSUE THREAD

**Issue by MhAllan** [dotnet/csharplang#8401]
IOnceEnumerable acts as IEnumerable if looped on it only once in the code, if more than one loops detected (more than one foreach, Count(), ... whatever) It should act as an Array. preventing the execution the method returning IEnumerable from scratch over and over. It will be had to detect I guess, but the feature is very useful, not always possible to track what developers do with IEnumerable, and performance can be degraded a lot when they loop on heavy IEnumerable creation more than once.

# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Motivation
[motivation]: prevent degrading the performance when developers mistakenly handle IEnumerable

## Detailed design
[design]: 
* Analyse the code and emit a class that implements IOnceEnumerable accordingly

## Drawbacks
[drawbacks]: deferred only once

## Alternatives
[alternatives]: ToArray() which people don't always pay attention to and hard to track in every code change

**Comment by yaakov-h** [comment]
What do you mean by the IEnumerable should "act as an array"?

**Comment by bernd5** [comment]
I think he means single evaluation of an enumerable. It would be great if we would have such a feature in the framework / runtime...


You can achieve this today by (e.g.:):

```csharp
using System;
using System.Collections;
using System.Collections.Generic;

var e = Enum();
var single = e.ForceSingleEvaluation();

foreach(var m in single)
{
    Console.WriteLine(m);
}
Console.WriteLine("---");
foreach (var m in single)
{
    Console.WriteLine(m);
}

IEnumerable<int> Enum()
{
    for (int i= 0; i < 10; i++)
    {
        Console.WriteLine($"Eval Enum({i})");
        yield return i;
    }
}

// Framework code

public static class IEnumerableExtensions
{
    public static IEnumerable<T> ForceSingleEvaluation<T>(this IEnumerable<T> e)
    {
        if (e is List<T> or T[] or SingleEvaluationEnumerable<T>)
        {
            return e;
        }
        return new SingleEvaluationEnumerable<T>(e);
    }
}

public class SingleEvaluationEnumerable<T>(IEnumerable<T> e) : IEnumerable<T>
{
    private List<T>? items;
    private IEnumerator<T>? nonEvaluated;

    public IEnumerator<T> GetEnumerator()
    {
        if (items is null)
        {
            items = new List<T>();
            nonEvaluated = e.GetEnumerator();
        }
        return new SingleEvaluationEnumerator<T>(items, nonEvaluated!);
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return this.GetEnumerator();
    }
}

public struct SingleEvaluationEnumerator<T>(List<T> items, IEnumerator<T> nonEvaluated, int idx = -1) : IEnumerator<T>
{
    public T Current
    {
        get
        {
            if (idx >= 0)
            {
                return items[idx];
            }
            throw new InvalidOperationException();
        }
    }

    object? IEnumerator.Current => this.Current;

    public void Dispose()
    {
    }

    public bool MoveNext()
    {
        var nextIdx = idx + 1;
        if (items.Count > nextIdx)
        {
            idx = nextIdx;
            return true;
        }
        if (nonEvaluated?.MoveNext() ?? false)
        {
            items.Add(nonEvaluated.Current);
            return MoveNext();
        }
        return false;
    }

    public void Reset()
    {
        idx = -1;
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Isn't this just ToList or ToArray?

**Comment by AmrAlSayed0** [comment]
@CyrusNajmabadi Not really. `IOnceEnumerable` won't iterate and store any element past the last element that was iterated by any `IEnumerator` that was created by that `IOnceEnumerable` instance. There is also the case of infinite `IEnumerable`s. A `FibonacciEnumerable` would only enumerate the numbers needed and if you call `FibonacciEnumerable.ElementAt` for an index that was previously enumerated it wouldn't recalculate it. Obviously, `ToList` and `ToArray` wouldn't work here.

**Comment by HaloFour** [comment]
Sounds like memoization to me.  I believe there is a `Memoize` extension method available in the [Interactive Extensions](https://github.com/dotnet/reactive/tree/main) for .NET.

**Comment by MhAllan** [comment]
@yaakov-h how do I know it should be opened as discussion, I went to issues and I found there is place to open proposal and I did so, I feel strange someone proposing something useful and just get closed! interesting!

@bernd5 Thank you for elaborating and for the perfect example. that is exactly what I am asking for. IEnumerable is evaluated everytime and that is not ok and hard to control by just reviewing PR's or rely on some IDE warning. IOnceEnumerable solves the problem. Thanks again @bernd5

**Comment by MhAllan** [comment]
> Isn't this just ToList or ToArray?

@CyrusNajmabadi yes and no. so when compiler sees that there is only one loop then it will act as notmal IEnumerable, but if more than one loop, it should implicitly consider to do ToArray. The compiler should also be able to find way to break at yield keyword even it is acting as Array. 

so for this code
```
IOnceEnumerable<int> Enumerate()
{
         DoSomething();
         yied return 1;
}
``` 

if I have one loop, that is normal IEnumerable, if there are more than one loop, the first loop should hit "DoSomething()" while the second loop should only hit "yield return 1";

**Comment by MhAllan** [comment]
> Sounds like memoization to me. I believe there is a `Memoize` extension method available in the [Interactive Extensions](https://github.com/dotnet/reactive/tree/main) for .NET.

@HaloFour not same, this is compiler generated instance that is created upon analysing the code, not just a buffer returning different enumerators

**Comment by theunrepentantgeek** [comment]
> how do I know it should be opened as discussion

It literally says so on the _New Issue_ screen:

![image](https://github.com/user-attachments/assets/19d45ae0-adad-43ec-8e8f-2117e57aee3f)

**Comment by HaloFour** [comment]
> not same, this is compiler generated instance that is created upon analysing the code, not just a buffer returning different enumerators

I don't see the difference, other that that one variation is infinitely more involved and infinitely less likely to happen.  Memoization solves this problem very nicely and requires zero work on the part of the language team.

**Comment by MhAllan** [comment]
@theunrepentantgeek I wrongly chose Language Specification.

**Comment by julealgon** [comment]
> @yaakov-h how do I know it should be opened as discussion, I went to issues and I found there is place to open proposal and I did so, I feel strange someone proposing something useful and just get closed! interesting!

@MhAllan , always read the readme/main page for the repository you are interacting with. Either it will describe contribution guidelines there (which is the case here) or there will be a separate, usually linked, `contributing.md` document with that information.

This is a general best practice and not specific to this repo or Microsoft repos.
